File: ml/predict.py
import sys, json, pickle, os
import numpy as np

BASE   = os.path.dirname(os.path.abspath(__file__))
MODELS = os.path.join(BASE,"models")

def load():
    def r(f): return pickle.load(open(os.path.join(MODELS,f),"rb"))
    def j(f): return json.load(open(os.path.join(MODELS,f)))
    return r("crop_model.pkl"),r("label_encoder.pkl"),j("rotation.json"),j("meta.json"),j("water.json"),j("pref_scores.json"),j("rythu_bandhu.json"),j("pmfby.json")

def soil_health(N,P,K,ph):
    ns = max(0,100-abs(N-100)*0.8)
    ps = max(0,100-abs(P-50)*1.2)
    ks = max(0,100-abs(K-50)*1.2)
    phs= max(0,100-abs(ph-7.0)*25)
    return round(min(100,ns*0.35+ps*0.25+ks*0.25+phs*0.15),1)

def npk_adequacy(N,P,K,crop):
    OPT={"Rice":(100,50,50),"Wheat":(125,65,50),"Maize":(100,55,45),
         "Chickpea":(28,60,35),"KidneyBeans":(35,65,45),"PigeonPeas":(28,65,35),
         "MothBeans":(20,35,20),"MungBean":(28,55,35),"Blackgram":(28,55,35),
         "Lentil":(20,55,35),"Pomegranate":(28,28,43),"Banana":(125,55,100),
         "Mango":(30,28,40),"Grapes":(30,28,40),"Watermelon":(100,40,50),
         "Muskmelon":(100,40,50),"Apple":(30,28,50),"Orange":(28,28,40),
         "Papaya":(70,40,50),"Coconut":(70,40,85),"Cotton":(130,65,55),"Jute":(90,55,50)}
    o=OPT.get(crop,(80,45,45))
    return {"N":round(min(100,(N/o[0])*100),1),"P":round(min(100,(P/o[1])*100),1),"K":round(min(100,(K/o[2])*100),1)}

def fert_reco(N,P,K,ph,crop):
    adq=npk_adequacy(N,P,K,crop)
    recs=[]
    if adq["N"]<70: recs.append(f"Apply Urea — Nitrogen deficiency ({adq['N']}% of optimal)")
    if adq["P"]<70: recs.append(f"Apply DAP — Phosphorus deficiency ({adq['P']}% of optimal)")
    if adq["K"]<70: recs.append(f"Apply MOP — Potassium deficiency ({adq['K']}% of optimal)")
    if ph<6.0: recs.append("Add agricultural lime to raise soil pH")
    elif ph>8.0: recs.append("Add gypsum or sulfur to lower soil pH")
    if not recs: recs.append("Soil nutrients are well-balanced for this crop ✅")
    return recs

def predict(inp):
    model,le,rotation,meta,water,pref_scores,rythu,pmfby = load()

    N=float(inp["N"]); P=float(inp["P"]); K=float(inp["K"]); ph=float(inp["ph"])
    temp=float(inp["temperature"]); hum=float(inp["humidity"]); rain=float(inp["rainfall"])
    prev=inp.get("previousCrop",""); pref=inp.get("preference","")
    water_avail=inp.get("waterAvailability","Medium")

    proba = model.predict_proba([[N,P,K,ph,temp,hum,rain]])[0].copy()

    # Preference boost (max 15%)
    for i,crop in enumerate(le.classes_):
        score = pref_scores.get(pref,{}).get(crop,0)
        proba[i] = min(1.0, proba[i] + score/100*0.15)

    # Water penalty
    wrank={"VeryLow":0,"Low":1,"Medium":2,"High":3}
    av=wrank.get(water_avail,2)
    for i,crop in enumerate(le.classes_):
        req=wrank.get(water.get(crop,"Medium"),2)
        if req>av: proba[i]=max(0,proba[i]-(req-av)*0.12)

    # Rotation — discourage repeating previous crop
    if prev in list(le.classes_):
        idx=list(le.classes_).index(prev)
        proba[idx]*=0.25

    proba/=proba.sum()
    top_idx=proba.argsort()[-3:][::-1]
    top3=[{"crop":le.classes_[i],"confidence":round(float(proba[i])*100,1)} for i in top_idx]
    best=top3[0]["crop"]

    sh=soil_health(N,P,K,ph)
    adq=npk_adequacy(N,P,K,best)
    fr=fert_reco(N,P,K,ph,best)

    return {
        "success":True,
        "topCrop":best,
        "confidence":top3[0]["confidence"],
        "top3":top3,
        "rotationSuggestions":rotation.get(best,[])[:3],
        "soilHealth":sh,
        "npkAdequacy":adq,
        "fertilizerRecommendation":fr,
        "waterRequirement":water.get(best,"Medium"),
        "rythuBandhu":best in rythu,
        "pmfbyCovered":best in pmfby,
        "modelAccuracy":meta["cv_accuracy"],
        "featureImportance":meta["feature_importance"],
    }

if __name__=="__main__":
    try:
        print(json.dumps(predict(json.loads(sys.argv[1]))))
    except Exception as e:
        print(json.dumps({"success":False,"error":str(e)}))
        sys.exit(1)

File: backend/routes/predict.js
const express = require("express");
const router  = express.Router();
const { spawn } = require("child_process");
const path    = require("path");

const ML = path.join(__dirname,"../../ml/predict.py");
const PY = process.env.PYTHON_PATH || "python";

const CROP_MAP = {
  Rice:"Rice",Wheat:"Wheat",Maize:"Maize",Chickpea:"Gram",
  KidneyBeans:"Rajma",PigeonPeas:"Tur",MothBeans:"Moth",
  MungBean:"Moong",Blackgram:"Urad",Lentil:"Masur",
  Pomegranate:"Pomegranate",Banana:"Banana",Mango:"Mango",
  Grapes:"Grapes",Watermelon:"Watermelon",Muskmelon:"Muskmelon",
  Apple:"Apple",Orange:"Orange",Papaya:"Papaya",
  Coconut:"Coconut",Cotton:"Cotton",Jute:"Jute",
};

async function fetchMandiPrice(crop, state, district) {
  const KEY = process.env.DATAGOV_KEY;
  const commodity = CROP_MAP[crop];
  if (!commodity || !KEY) return null;

  const base = `https://api.data.gov.in/resource/9ef84268-d588-465a-a308-a864a43d0070?api-key=${KEY}&format=json&limit=10&filters[commodity]=${encodeURIComponent(commodity)}`;

  const tryFetch = async (url) => {
    try {
      const r  = await fetch(url, { signal: AbortSignal.timeout(8000) });
      const d  = await r.json();
      return (d.records||[]).map(rec=>({
        market:   rec.market        || "—",
        district: rec.district      || "—",
        state:    rec.state         || "—",
        min:      parseFloat(rec.min_price   || 0),
        max:      parseFloat(rec.max_price   || 0),
        modal:    parseFloat(rec.modal_price || 0),
        date:     rec.arrival_date  || "—",
      })).filter(r=>r.modal>0);
    } catch { return []; }
  };

  // Try district → state → national (in parallel where possible)
  let prices = [];
  if (state === "Telangana" || state === "Andhra Pradesh") {
    // Prioritise Telangana/AP districts
    if (district) prices = await tryFetch(base+`&filters[state.keyword]=${encodeURIComponent(state)}&filters[district]=${encodeURIComponent(district)}`);
    if (!prices.length) prices = await tryFetch(base+`&filters[state.keyword]=${encodeURIComponent(state)}`);
  }
  if (!prices.length && district) prices = await tryFetch(base+`&filters[district]=${encodeURIComponent(district)}`);
  if (!prices.length && state)    prices = await tryFetch(base+`&filters[state.keyword]=${encodeURIComponent(state)}`);
  if (!prices.length)             prices = await tryFetch(base);
  if (!prices.length) return null;

  const modals = prices.map(p=>p.modal);
  return {
    prices: prices.slice(0,5),
    avg:    Math.round(modals.reduce((a,b)=>a+b,0)/modals.length),
    min:    Math.min(...prices.map(p=>p.min)),
    max:    Math.max(...prices.map(p=>p.max)),
    count:  prices.length,
  };
}

function runML(inputData) {
  return new Promise((resolve, reject) => {
    let out="", err="";
    const py = spawn(PY, [ML, JSON.stringify(inputData)]);
    py.stdout.on("data", d=>out+=d.toString());
    py.stderr.on("data", d=>err+=d.toString());
    py.on("close", code=>{
      if (code!==0) return reject(new Error(err));
      try { resolve(JSON.parse(out.trim())); }
      catch(e) { reject(new Error("Parse error: "+out)); }
    });
  });
}

router.post("/", async (req,res) => {
  const { N,P,K,ph,temperature,humidity,rainfall,
          previousCrop,soilType,season,preference,
          waterAvailability,district,state,lat,lon,mode } = req.body;

  if ([N,P,K,ph,temperature,humidity,rainfall].some(v=>v===undefined||v===""))
    return res.status(400).json({success:false,error:"All soil/weather fields required."});

  try {
    // Run ML and fetch mandi price SIMULTANEOUSLY
    const [mlResult, mandiData] = await Promise.all([
      runML({N,P,K,ph,temperature,humidity,rainfall,
             previousCrop:previousCrop||"",
             preference:preference||"",
             waterAvailability:waterAvailability||"Medium"}),
      fetchMandiPrice(null, state||"", district||"") // will re-fetch after ML for correct crop
    ]);

    // Now fetch mandi for the actual recommended crop
    const mandiForCrop = await fetchMandiPrice(mlResult.topCrop, state||"", district||"");

    const result = {
      ...mlResult,
      mandi: mandiForCrop,
      inputs:{ N,P,K,ph,temperature,humidity,rainfall,
               previousCrop:previousCrop||"",soilType:soilType||"",
               season:season||"",preference:preference||"",
               waterAvailability:waterAvailability||"Medium",
               district:district||"",state:state||"",
               lat,lon,mode:mode||"simple" }
    };

    // Save to DB async — don't wait for it
    try {
      const History = require("../models/History");
      new History({
        inputs:result.inputs, topCrop:result.topCrop,
        confidence:result.confidence, top3:result.top3,
        rotationSuggestions:result.rotationSuggestions,
        soilHealth:result.soilHealth, npkAdequacy:result.npkAdequacy,
        fertilizerRecommendation:result.fertilizerRecommendation,
        waterRequirement:result.waterRequirement,
        rythuBandhu:result.rythuBandhu, pmfbyCovered:result.pmfbyCovered,
        district:district||"", state:state||"",
        preference:preference||"", mode:mode||"simple",
      }).save().catch(()=>{});
    } catch {}

    res.json(result);
  } catch(e) {
    res.status(500).json({success:false,error:e.message});
  }
});

module.exports = router;
